### PR TITLE
(maint) Update docker alpine 3.11

### PR DESCRIPTION
**Pull request by Iristyle**
Builds on #1901 (which actually can't be merged since Travis has problems with Alpine 3.10)

- Fixes minor bug in Alpine builds
- Tracks what's necessary to upgrade Alpine to 3.10 from 3.9 (removal of boost-signals, addition of boost-locale)
- Updates to Alpine 3.11 by ... 
- Requires pinning cmake back to 3.10 repo, as the 3.11 version of cmake
  crashes if it doesn't have libstdc++-9.2.0-r4, and gcc 6 / g++ 6
  require the older libstdc++ 6 series

- Similarly, pin back boost-dev=1.69.0-r4 yaml-cpp-dev=0.6.2-r2 for the
  same reason as the linking process will fail with errors like:

 #11 175.7 /usr/lib/gcc/x86_64-alpine-linux-musl/6.4.0/../../../../x86_64-alpine-linux-musl/bin/ld: /usr/lib/libboost_locale.so.1.71.0: undefined reference to `std::__cxx11::basic_ostringstream<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::basic_ostringstream()'
 #11 175.7 /usr/lib/gcc/x86_64-alpine-linux-musl/6.4.0/../../../../x86_64-alpine-linux-musl/bin/ld: /usr/lib/libboost_locale.so.1.71.0: undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'
 #11 175.7 /usr/lib/gcc/x86_64-alpine-linux-musl/6.4.0/../../../../x86_64-alpine-linux-musl/bin/ld: /usr/lib/libboost_locale.so.1.71.0: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'
 #11 175.7 collect2: error: ld returned 1 exit status

 #29 467.9 /usr/lib/gcc/x86_64-alpine-linux-musl/6.4.0/../../../../x86_64-alpine-linux-musl/bin/ld: /usr/lib/gcc/x86_64-alpine-linux-musl/6.4.0/../../../../lib/libyaml-cpp.so: undefined reference to `std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::basic_stringstream()'

 These both require the newer libstdc++ for ABI compatibility, but we're
 forced to use the older libstdc++ due to required compiler toolchain.

- In the runtime container, also pull binary packages from the 3.10
  repository for the relevant boost libraries and yaml-cpp. Add the
  boost-locale package.

- Upgrades Ruby to 2.6, which requires a change from --no-ri --no-rdoc
  to --no-document when installing gems.

### CODE REVIEW

**Comment by Iristyle** [comment]
It looks like the image size has doubled from about 100MB to 200MB on my local machine going from 3.10 to 3.11. It looks like this could be due to multiple copies of Ruby being installed, in addition to Python 2.7 and Python 3.2 (though I'm not yet sure how that has happened). Digging in!

File: docker/puppet-agent-alpine/Dockerfile [hunk 1]
```diff
@@ -180,36 +183,43 @@ LABEL org.label-schema.maintainer="Puppet Release Team <release@puppet.com>" \
...
+    boost-log=1.69.0-r4 \
+    boost-math=1.69.0-r4 \
+    boost-prg_exec_monitor=1.69.0-r4 \
+    boost-program_options=1.69.0-r4 \
+    boost-random=1.69.0-r4 \
+    boost-regex=1.69.0-r4 \
+    boost-serialization=1.69.0-r4 \
+    boost-system=1.69.0-r4 \
+    boost-thread=1.69.0-r4 \
+    boost-unit_test_framework=1.69.0-r4 \
```
Comment (Iristyle) [inline docker/puppet-agent-alpine/Dockerfile:207]: We might still be carrying a few unnecessary boost libraries, but leaving the prior art for now.

**Comment by Iristyle** [comment]
Current diff between a 6.15.0 candidate container based on 3.11 and a 3.9 based 6.14.0

```diff
--- from_file
+++ 6.14.0
@@ -1,4 +1,4 @@
+6.15.0
-6.14.0
-boost
+boost-locale
+boost-log
-boost-signals
+ca-certificates-cacert
+fstrm
+icu-libs
+keyutils-libs
+krb5-conf
+krb5-libs
+libcom_err
+libcrypto1.1
+libprotobuf
+libprotoc
+libssl1.1
-libressl2.7-libcrypto
-libressl2.7-libssl
-libressl2.7-libtls
-libssh2
+libtls-standalone
+libverto
-ncurses-terminfo
+protobuf-c
+ruby-etc
+ruby-multi_json
+virt-what
+xz-libs
+zstd-libs
```

**Comment by hdeadman** [comment]
> want to revisit bumping up to Alpine latest (3.13) since 3.9 stopped publishing a year ago and is probably very out of date on security updates.

I can confirm that it is very out of date on security updates. This is a scan of the latest. 
```
Vulnerabilities
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
|      CVE       | SEVERITY | CVSS |  PACKAGE  |     VERSION      |          STATUS           | PUBLISHED  | DISCOVERED |                    DESCRIPTION                     |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-3518  | high     | 8.80 | libxml2   | 2.9.9-r3         | fixed in 2.9.12-r1        | 52 days    | < 1 hour   | There\'s a flaw in libxml2 in versions before      |
|                |          |      |           |                  | 52 days ago               |            |            | 2.9.11. An attacker who is able to submit a        |
|                |          |      |           |                  |                           |            |            | crafted file to be processed by an application     |
|                |          |      |           |                  |                           |            |            | linked with l...                                   |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-3517  | high     | 8.60 | libxml2   | 2.9.9-r3         | fixed in 2.9.12-r1        | 51 days    | < 1 hour   | There is a flaw in the xml entity encoding         |
|                |          |      |           |                  | 51 days ago               |            |            | functionality of libxml2 in versions before        |
|                |          |      |           |                  |                           |            |            | 2.9.11. An attacker who is able to supply a        |
|                |          |      |           |                  |                           |            |            | crafted file to be ...                             |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2019-18276 | high     | 7.80 | bash      | 4.4.19-r1        | fixed in 5.0.11-r1        | > 1 years  | < 1 hour   | An issue was discovered in disable_priv_mode       |
|                |          |      |           |                  | > 1 years ago             |            |            | in shell.c in GNU Bash through 5.0 patch 11. By    |
|                |          |      |           |                  |                           |            |            | default, if Bash is run with its effective UID not |
|                |          |      |           |                  |                           |            |            | equal ...                                          |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-30139 | high     | 7.50 | apk-tools | 2.10.3-r1        | fixed in 2.12.5-r0        | 78 days    | < 1 hour   | In Alpine Linux apk-tools before 2.12.5, the       |
|                |          |      |           |                  | 78 days ago               |            |            | tarball parser allows a buffer overflow and crash. |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-28965 | high     | 7.50 | ruby      | 2.5.8-r0         | fixed in 2.6.7-r0         | 79 days    | < 1 hour   | The REXML gem before 3.2.5 in Ruby before 2.6.7,   |
|                |          |      |           |                  | 79 days ago               |            |            | 2.7.x before 2.7.3, and 3.x before 3.0.1 does      |
|                |          |      |           |                  |                           |            |            | not properly address XML round-trip issues. An     |
|                |          |      |           |                  |                           |            |            | incorrec...                                        |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-23840 | high     | 7.50 | openssl   | 1.1.1g-r0        | fixed in 1.1.1j-r0        | > 4 months | < 1 hour   | Calls to EVP_CipherUpdate, EVP_EncryptUpdate and   |
|                |          |      |           |                  | > 4 months ago            |            |            | EVP_DecryptUpdate may overflow the output length   |
|                |          |      |           |                  |                           |            |            | argument in some cases where the input length is   |
|                |          |      |           |                  |                           |            |            | clo...                                             |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2020-8231  | high     | 7.50 | curl      | 7.64.0-r5        | fixed in 7.66.0-r5        | > 6 months | < 1 hour   | Due to use of a dangling pointer, libcurl 7.29.0   |
|                |          |      |           |                  | > 6 months ago            |            |            | through 7.71.1 can use the wrong connection when   |
|                |          |      |           |                  |                           |            |            | sending data.                                      |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2020-25613 | high     | 7.50 | ruby      | 2.5.8-r0         | fixed in 2.5.9-r0         | > 9 months | < 1 hour   | An issue was discovered in Ruby through 2.5.8,     |
|                |          |      |           |                  | > 9 months ago            |            |            | 2.6.x through 2.6.6, and 2.7.x through 2.7.1.      |
|                |          |      |           |                  |                           |            |            | WEBrick, a simple HTTP server bundled with Ruby,   |
|                |          |      |           |                  |                           |            |            | had not ...                                        |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2020-10663 | high     | 7.50 | json      | 2.1.0            |                           | > 1 years  | < 1 hour   | The JSON gem through 2.2.0 for Ruby, as used in    |
|                |          |      |           |                  |                           |            |            | Ruby 2.4 through 2.4.9, 2.5 through 2.5.7, and     |
|                |          |      |           |                  |                           |            |            | 2.6 through 2.6.5, has an Unsafe Object Creation   |
|                |          |      |           |                  |                           |            |            | Vulner...                                          |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2019-6470  | high     | 7.50 | bind      | 9.12.4_p2-r0     |                           | > 1 years  | < 1 hour   | There had existed in one of the ISC BIND libraries |
|                |          |      |           |                  |                           |            |            | a bug in a function that was used by dhcpd when    |
|                |          |      |           |                  |                           |            |            | operating in DHCPv6 mode. There was also a bug in  |
|                |          |      |           |                  |                           |            |            | d...                                               |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2019-15847 | high     | 7.50 | gcc       | 8.3.0-r0         | fixed in 9.3.0-r0         | > 1 years  | < 1 hour   | The POWER9 backend in GNU Compiler Collection      |
|                |          |      |           |                  | > 1 years ago             |            |            | (GCC) before version 10 could optimize multiple    |
|                |          |      |           |                  |                           |            |            | calls of the __builtin_darn intrinsic into a       |
|                |          |      |           |                  |                           |            |            | single call...                                     |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-3450  | high     | 7.40 | openssl   | 1.1.1g-r0        | fixed in 1.1.1k-r0        | > 3 months | < 1 hour   | The X509_V_FLAG_X509_STRICT flag enables           |
|                |          |      |           |                  | > 3 months ago            |            |            | additional security checks of the certificates     |
|                |          |      |           |                  |                           |            |            | present in a certificate chain. It is not set by   |
|                |          |      |           |                  |                           |            |            | default. Star...                                   |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-3537  | medium   | 5.90 | libxml2   | 2.9.9-r3         | fixed in 2.9.12-r1        | 55 days    | < 1 hour   | A vulnerability found in libxml2 in versions       |
|                |          |      |           |                  | 55 days ago               |            |            | before 2.9.11 shows that it did not propagate      |
|                |          |      |           |                  |                           |            |            | errors while parsing XML mixed content, causing a  |
|                |          |      |           |                  |                           |            |            | NULL dere...                                       |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-3449  | medium   | 5.90 | openssl   | 1.1.1g-r0        | fixed in 1.1.1k-r0        | > 3 months | < 1 hour   | An OpenSSL TLS server may crash if sent a          |
|                |          |      |           |                  | > 3 months ago            |            |            | maliciously crafted renegotiation ClientHello      |
|                |          |      |           |                  |                           |            |            | message from a client. If a TLSv1.2 renegotiation  |
|                |          |      |           |                  |                           |            |            | ClientHello ...                                    |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-23841 | medium   | 5.90 | openssl   | 1.1.1g-r0        | fixed in 1.1.1j-r0        | > 4 months | < 1 hour   | The OpenSSL public API function                    |
|                |          |      |           |                  | > 4 months ago            |            |            | X509_issuer_and_serial_hash() attempts to create   |
|                |          |      |           |                  |                           |            |            | a unique hash value based on the issuer and serial |
|                |          |      |           |                  |                           |            |            | number data contai...                              |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2020-1971  | medium   | 5.90 | openssl   | 1.1.1g-r0        | fixed in 1.1.1i-r0        | > 7 months | < 1 hour   | The X.509 GeneralName type is a generic type for   |
|                |          |      |           |                  | > 7 months ago            |            |            | representing different types of names. One of      |
|                |          |      |           |                  |                           |            |            | those name types is known as EDIPartyName. OpenSSL |
|                |          |      |           |                  |                           |            |            | prov...                                            |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2020-28928 | medium   | 5.50 | musl      | 1.1.20-r5        | fixed in 1.1.20-r6        | > 7 months | < 1 hour   | In musl libc through 1.2.1, wcsnrtombs mishandles  |
|                |          |      |           |                  | > 7 months ago            |            |            | particular combinations of destination buffer size |
|                |          |      |           |                  |                           |            |            | and source character limit, as demonstrated by     |
|                |          |      |           |                  |                           |            |            | an...                                              |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2019-17595 | medium   | 5.40 | ncurses   | 6.1_p20190105-r0 | fixed in 6.2_p20200523-r0 | > 1 years  | < 1 hour   | There is a heap-based buffer over-read in the      |
|                |          |      |           |                  | > 1 years ago             |            |            | fmt_entry function in tinfo/comp_hash.c in the     |
|                |          |      |           |                  |                           |            |            | terminfo library in ncurses before 6.1-20191012.   |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-22897 | medium   | 5.30 | curl      | 7.64.0-r5        | fixed in 7.77.0-r0        | 27 days    | < 1 hour   | curl 7.61.0 through 7.76.1 suffers from exposure   |
|                |          |      |           |                  | 27 days ago               |            |            | of data element to wrong session due to a mistake  |
|                |          |      |           |                  |                           |            |            | in the code for CURLOPT_SSL_CIPHER_LIST when       |
|                |          |      |           |                  |                           |            |            | libcur...                                          |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2019-17594 | medium   | 5.30 | ncurses   | 6.1_p20190105-r0 | fixed in 6.2_p20200523-r0 | > 1 years  | < 1 hour   | There is a heap-based buffer over-read in the      |
|                |          |      |           |                  | > 1 years ago             |            |            | _nc_find_entry function in tinfo/comp_hash.c       |
|                |          |      |           |                  |                           |            |            | in the terminfo library in ncurses before          |
|                |          |      |           |                  |                           |            |            | 6.1-20191012.                                      |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-23839 | low      | 3.70 | openssl   | 1.1.1g-r0        | fixed in 1.1.1j-r0        | > 4 months | < 1 hour   | OpenSSL 1.0.2 supports SSLv2. If a client          |
|                |          |      |           |                  | > 4 months ago            |            |            | attempts to negotiate SSLv2 with a server that is  |
|                |          |      |           |                  |                           |            |            | configured to support both SSLv2 and more recent   |
|                |          |      |           |                  |                           |            |            | SSL and T...                                       |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2020-8284  | low      | 3.70 | curl      | 7.64.0-r5        | fixed in 7.77.0-r0        | > 6 months | < 1 hour   | A malicious server can use the FTP PASV response   |
|                |          |      |           |                  | > 6 months ago            |            |            | to trick curl 7.73.0 and earlier into connecting   |
|                |          |      |           |                  |                           |            |            | back to a given IP address and port, and this way  |
|                |          |      |           |                  |                           |            |            | po...                                              |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+
| CVE-2021-22898 | low      | 3.10 | curl      | 7.64.0-r5        | fixed in 7.77.0-r0        | 27 days    | < 1 hour   | curl 7.7 through 7.76.1 suffers from an            |
|                |          |      |           |                  | 27 days ago               |            |            | information disclosure when the `-t` command       |
|                |          |      |           |                  |                           |            |            | line option, known as `CURLOPT_TELNETOPTIONS` in   |
|                |          |      |           |                  |                           |            |            | libcurl, is used...                                |
+----------------+----------+------+-----------+------------------+---------------------------+------------+------------+----------------------------------------------------+```

**Comment by Iristyle** [comment]
Yeah I think we should discuss internally, but I'd be in favor of just discontinuing publishing the alpine agent altogether.

/cc @underscorgan @binford2k

**Comment by joshcooper** [comment]
We don't have the bandwidth to continue supporting alpine based images, so I'm going to close this and remove the alpine specific dockerfiles, see PR #2216.

**Pull request closed without merging**